File: src/danbooru/rule34_client.py
import random

from danbooru.danbooru_client import DanbooruClient


class Rule34Client(DanbooruClient):
    def _get_index_params(self):
        page = random.randint(1, 2000)
        return {"page": "dapi", "s": "post", "q": "index", "json": "1", "limit": "100", "pid": str(page)}

    _title = "Rule 34"
    _index_url = "https://rule34.xxx/index.php"

    def _is_explicit(self, post):
        rating = post["rating"]
        if rating is None or len(rating) <= 0:
            return False

        if rating.lower() == "explicit":
            return True
        else:
            return False

    def _get_image_url(self, post):
        return "https://rule34.xxx/images/" + post["directory"] + "/" + post["image"]

    def _get_full_post_url(self, post):
        return "https://rule34.xxx/index.php?page=post&s=view&id=" + str(post["id"])

    def _get_tag_list(self, post):
        return post["tags"].split()

File: src/main/kotlin/com/github/hannesbraun/katarina/modules/danbooru/Rule34Client.kt
package com.github.hannesbraun.katarina.modules.danbooru

import com.google.gson.JsonObject
import kotlinx.coroutines.CoroutineScope
import kotlin.random.Random

class Rule34Client(private val scope: CoroutineScope) : DanbooruClient(scope) {
    override val name = "Rule 34"
    override val indexUrl = "https://rule34.xxx/index.php"

    override fun getIndexParameters(): List<Pair<String, String>> {
        val pid = Random.nextInt(1, 2001)
        return listOf(
            Pair("page", "dapi"),
            Pair("s", "post"),
            Pair("q", "index"),
            Pair("json", "1"),
            Pair("limit", postBufferSize.toString()),
            Pair("pid", pid.toString())
        )
    }

    override fun generatePost(
        id: String, fileUrl: String, rating: String, score: String, tagList: List<String>
    ) = Rule34Post(id, fileUrl, rating, score, tagList)

    override fun getTagListFromJson(post: JsonObject): List<String> = post.getString("tags").split(" ")
    override fun getFileUrlFromJson(post: JsonObject) =
        "https://rule34.xxx/images/${post.getString("directory")}/${post.getString("image")}"
}

class Rule34Post(id: String, fileUrl: String, rating: String, score: String, tagList: List<String>) : DanbooruPost(
    id,
    fileUrl,
    rating,
    score,
    tagList
) {
    override fun getFullPostUrl() = "https://rule34.xxx/index.php?page=post&s=view&id=$id"
    override fun isExplicit() = rating.toLowerCase() == "explicit"
}